src/main/java/moe/nightfall/vic/integratedcircuits/cp/part/PartWire.java

```diff
@@ -35,10 +35,17 @@ public boolean getOutputToSide(Vec2 pos, ICircuit parent, ForgeDirection side) {
 
 	@Override
 	@SideOnly(Side.CLIENT)
+	// This renders BOTH WIRES AND TUNNELS
 	public void renderPart(Vec2 pos, ICircuit parent, double x, double y, CircuitPartRenderer.EnumRenderType type) {
 		int color = this.getColor(pos, parent);
+		boolean tun = this instanceof PartTunnel;
```

Comment:
Properly using OOP is hell. I mean designing a proper class hierarchy.

I know that this check is actually `this.canDrawAsStraightLine()`, plus `PartTunnel` should render background before calling `super.renderPart`.

Oh wait! Call super is antipattern too! So this should actually be a `renderWireLike` method (probably static) that receives `allowStraightLine` flag among other arguments, and it should be called by both wire and tunnel renderers.

And what about [`if (selectedPart.getPart() instanceof PartWire)`](https://github.com/Victorious3/Integrated-Circuits/pull/179/files#diff-31f7bf9fb3725fbdc5aa9ac39d740e68L54)? Surely it should be a virtual `allowDragPlacement()` method of `CircuitPart` class.
Oh, and maybe `PartWire` should decide how it is drawn during placement and not `CadGui`? I mean `CircuitPart.renderDuringPlacement(from, to, ...)` and `CircuitPart.doPlace(from, to, ...)` methods. Overriding these methods will allow, for example, to rotate gates during placement, or set delay of timers proportional to distance, or to place pair of connected tunnels.
Or even better, these "placers" should not be implicitly tied to gates they place. This will allow to implement "connect 2 wires to turn them into tunnels" thing cleanly.

In general, hardcoding is bad, but avoiding hardcoding takes time and decreases runtime performance, while absolutely no hardcoding means infinite code. I do not mean that I reject your suggestion to "find an OOP way of doing this", just that I see no practical reason to not simply use `part isinstance PartTunnel` unless this check has obvious potential of becoming something more complex.

The code base has a ton of cleanup potential, and would likely benefit from more comments and documentation. But I guess no one has time to refactor it.
Also, I have been (un)lucky to experience how perfectionism can be a *really* bad thing in terms of time consumption and project completion, so my opinion is that result is what matters. And the result is that the code works and is more or less readable.